vaadin-flow-components-shared-parent/vaadin-flow-components-base/src/main/java/com/vaadin/flow/component/shared/internal/DisableOnClickController.java

```diff
@@ -0,0 +1,135 @@
+        implements Serializable {
+
+    private final C component;
+    private boolean disableOnClick = false;
+    private PendingJavaScriptResult initDisableOnClick;
+
+    /**
+     * Creates a new controller for the given component.
+     * <p>
+     * The component must implement {@link HasEnabled} and
+     * {@link ClickNotifier}.
+     *
+     * @param component
+     *            the component to control, not {@code null}
+     */
+    public DisableOnClickController(C component) {
+        this.component = Objects.requireNonNull(component);
+        component.addClickListener(itemClickEvent -> {
+            if (isDisableOnClick()) {
+                component.setEnabled(false);
+            }
+        });
+        component.addAttachListener(aAttachEvent -> {
+            if (isDisableOnClick()) {
+                initDisableOnClick();
+            }
+        });
+    }
+
+    /**
+     * Sets whether the component should be disabled when clicked.
+     * <p>
+     * When set to {@code true}, the component will be immediately disabled on
+     * the client-side when clicked, preventing further clicks until re-enabled
+     * from the server-side.
+     *
+     * @param disableOnClick
+     *            whether the component should be disabled when clicked
+     */
+    public void setDisableOnClick(boolean disableOnClick) {
+        this.disableOnClick = disableOnClick;
+        if (disableOnClick) {
+            component.getElement().setAttribute("disableOnClick", "true");
+            initDisableOnClick();
+        } else {
+            component.getElement().removeAttribute("disableOnClick");
+        }
+    }
+
+    /**
+     * Gets whether the component is set to be disabled when clicked.
+     *
+     * @return whether the component is set to be disabled on click
+     */
+    public boolean isDisableOnClick() {
+        return disableOnClick;
+    }
+
+    /**
+     * Forces the client-side component's {@code disabled} property to be
+     * updated immediately.
+     * <p>
+     * This method should be called from the component's
+     * {@link HasEnabled#setEnabled} method.
+     *
+     * @param enabled
+     *            value to set
+     */
+    public void onSetEnabled(boolean enabled) {
+        // If the component is then disabled and re-enabled during the same
+        // round trip, Flow will not detect any changes and the client side
+        // component would not be enabled again.
+        component.getElement().executeJs("this.disabled = $0", !enabled);
+    }
+
+    /**
+     * Initialize client side disabling so disabled if immediate on click even
+     * if server-side handling takes some time.
+     */
+    private void initDisableOnClick() {
+        if (initDisableOnClick == null) {
+            initDisableOnClick = component.getElement().executeJs("""
+                    const el = $0;
+                    if (el) {
```

Comment:
In what situations, can `el` be null? It seems like the same logic could be expressed as just:

```java
component.getElement()
    .executeJs("window.Vaadin.Flow.disableOnClick.initDisableOnClick(this)");

```

Reply:
That was an error I encountered while testing context menu with attach/detach. However, after refactoring, there are no more errors. Therefore, I reverted this to the original.